pkg/provisioner/ironic/ironic.go

```diff
@@ -1336,6 +1340,34 @@ func (p *ironicProvisioner) Delete() (result provisioner.Result, err error) {
 		return p.setMaintenanceFlag(ironicNode, true)
 	}
 
+	if ironicNode.PowerState == powerOn {
+		p.log.Info("host is being powered off before removal")
+		provResult, err := p.hardPowerOff()
+		if err != nil {
+			switch err.(type) {
+			case HostLockedError:
+				result.RequeueAfter = powerRequeueDelay
+				result.Dirty = true
+				return result, err
+			default:
+				if result.ErrorCount < maxErrorRetryCount {
```

Comment:
I don't think this is going to do what you expect. This data structure is not persisted anywhere, so the count will never grow higher than 1, and it will never be more than 0 at this point because we will always have a fresh result structure.

I think instead of adding an ErrorCount value to the result structure, you want to look at the host.Status.ErrorCount. That is maintained by the controller when it sees an error from one of the methods while it is reconciling the host.

Maybe @andfasano can confirm that would work.

Reply:
Sorry for the late answer, but @dhellmann is correct, as the business logic should be based on the `host.Status.ErrorCount`. The provisioner then can decide how to proceed by using one of the semantic methods introduced by @zaneb in the [result.go](https://github.com/metal3-io/baremetal-operator/blob/ea587bb61a93547cebd1483be5852d1104b35d33/pkg/provisioner/ironic/result.go) - they improve also the code readability.

Reply:
Yeah, I had wrongly understood result to be preserved across subsequent calls to the same method. I also see that the code around error handling has evolved considerably. So, starting afresh to align with these changes.